### DISCUSSION TITLE
Metal support in NanoVDB?

### DISCUSSION THREAD

**Discussion by monroewilliams** [AcademySoftwareFoundation/openvdb#1773] [General]
I'm looking into the feasibility of using NanoVDB for rendering in a project which will run on several platforms, including iOS.
To make this work well, I would need to use NanoVDB with the Metal rendering API, and I don't currently see any support for that in the master branch.

Searching around a bit, I do see this PR in the Blender project, which includes patches to NanoVDB that add Metal support (which according to the person who opened the PR were submitted by someone at Apple). 

[Cycles: add NanoVDB support for Metal on Apple Silicon](https://projects.blender.org/blender/blender/pulls/104837)

Obviously this is not my work, so it wouldn't be appropriate for me to bring these patches in as a PR, but I would like to know if there are any discussions ongoing about incorporating these changes (or something similar) into mainline OpenVDB. 

(For my own purposes I can of course use patched sources or a private fork, but it's always nice to see these things pushed upstream as appropriate.)

**Comment by kmuseth** [comment]
I had a quick glance at the PR you’re referring to and I don’t see any changes to NanoVDB.h itself, only cyclone and some external (interpolation) methods. I don’t have any experience with Metal (or plans to support it), but I’m certainly open to the idea of adding support for it if someone can either submit the required changes or explain to me why the current version isn’t compatible with Metal.Thanks,KenOn Mar 6, 2024, at 12:00 PM, Monroe Williams ***@***.***> wrote:﻿
I'm looking into the feasibility of using NanoVDB for rendering in a project which will run on several platforms, including iOS.
To make this work well, I would need to use NanoVDB with the Metal rendering API, and I don't currently see any support for that in the master branch.
Searching around a bit, I do see this PR in the Blender project, which includes patches to NanoVDB that add Metal support (which according to the person who opened the PR were submitted by someone at Apple).
https://projects.blender.org/blender/blender/pulls/104590
Obviously this is not my work, so it wouldn't be appropriate for me to bring these patches in as a PR, but I would like to know if there are any discussions ongoing about incorporating these changes into mainline OpenVDB.
(For my own purposes I can of course use patched sources or a private fork, but it's always nice to see these things pushed upstream as appropriate.)

**Reply by monroewilliams** [reply]
The NanoVDB changes are in a patch file, which the PR diff doesn't display due to its size. 

https://projects.blender.org/blender/blender/src/commit/431ce537e6032b5d1fbe5240cc1999e1c34ff3db/build_files/build_environment/patches/openvdb_metal.diff

**Reply by monroewilliams** [reply]
I haven't fully evaluated the patch yet myself (as the project I'm working on is still in the "evaluating feasibility" stage).
At least part of what the patch does is to add an additional `__KERNEL_METAL__` case to NanoVDB.h which is a peer to the existing `__CUDACC_RTC__` case.  
Looking at it further, it seems as though the patch doesn't apply cleanly to the current tip of the master branch in OpenVDB. (It's not clear which version of OpenVDB they're building against, and that PR was originally created over a year ago, so it may be stale.) Possibly reaching out to the original creator would be in order.

I'm pretty familiar with Metal, but not very familiar with OpenVDB/NanoVDB (yet). I would expect an additional case may be needed in PNanoVDB.h for the Metal shading language (which is conceptually equivalent to, but distinct from, GLSL and HLSL), but this patch doesn't seem to add that.
